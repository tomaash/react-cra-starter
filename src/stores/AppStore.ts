import { observable } from 'mobx'
import axios from 'axios'
import { History, HistoryListenerParameter } from '@reach/router'

interface User {
  id: number
  first_name: string
  last_name: string
  avatar: string
}

interface ApiResponse {
  page: 1
  per_page: 3
  total: 12
  total_pages: 4
}

interface UserResponse extends ApiResponse {
  data: User[]
}

export class AppStore {
  @observable users: User[] = []
  @observable currentRoute: HistoryListenerParameter

  setupHistory(history: History) {
    history.listen(state => {
      console.log(state)
      this.currentRoute = state
    })
  }

  loadUsers = async () => {
    const res = await axios.get('https://reqres.in/api/users')
    this.users = res.data.data
  }
}
