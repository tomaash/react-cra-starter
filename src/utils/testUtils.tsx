import React from 'react'
import ReactDOMServer from 'react-dom/server'
const path = require('path')

export const squashHtmlTags = str => str.replace(/(<[^>]+>)+/g, '\n')

export const renderText = element =>
  squashHtmlTags(ReactDOMServer.renderToStaticMarkup(element))

export const startPolly = dirname => {
  const { Polly } = require('@pollyjs/core')
  const { setupPolly } = require('setup-polly-jest')
  const FSPersister = require('@pollyjs/persister-fs')
  const XHRAdapter = require('@pollyjs/adapter-xhr')

  Polly.register(XHRAdapter)
  Polly.register(FSPersister)

  setupPolly({
    adapters: ['xhr'],
    recordIfMissing: true,
    persister: 'fs',
    persisterOptions: {
      fs: {
        recordingsDir: path.resolve(dirname, '__recordings__')
      }
    }
  })
}

export const sleep = m => new Promise(r => setTimeout(r, m))
