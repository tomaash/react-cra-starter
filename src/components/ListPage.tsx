import React, { Component } from 'react'
import { observer, inject } from 'mobx-react'
import { observable, toJS } from 'mobx'
import {
  Table,
  TableBody,
  TableCell,
  TableHead,
  TableRow
} from '@material-ui/core'
import { AppStore } from '~/stores/AppStore'
import { RouteComponentProps } from '@reach/router'

interface ListPageProps extends RouteComponentProps {
  appStore?: AppStore
}

@inject('appStore')
@observer
export class ListPage extends Component<ListPageProps> {
  componentDidMount() {
    this.props.appStore.loadUsers()
  }
  render() {
    const { users } = this.props.appStore
    return (
      <>
        <Table>
          <TableHead>
            <TableRow>
              <TableCell>First Name</TableCell>
              <TableCell>Last Name</TableCell>
            </TableRow>
          </TableHead>
          <TableBody>
            {users.map(item => {
              return (
                <TableRow key={item.id}>
                  <TableCell>{item.first_name}</TableCell>
                  <TableCell>{item.last_name}</TableCell>
                </TableRow>
              )
            })}
          </TableBody>
        </Table>
      </>
    )
  }
}
