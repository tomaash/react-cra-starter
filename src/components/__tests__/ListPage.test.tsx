import React from 'react'
import { mount } from 'enzyme'
import { squashHtmlTags, renderText, startPolly } from '~/utils/testUtils'
import { ListPage } from '../ListPage'
import { AppStore } from '~/stores/AppStore'
import { Provider } from 'mobx-react'
import { when } from 'mobx'

describe('<ListPage />', () => {
  startPolly(__dirname)

  it('renders component', () => {
    const componentSetup = (
      <Provider appStore={new AppStore()}>
        <ListPage />
      </Provider>
    )

    const text = renderText(componentSetup)
    expect(squashHtmlTags(text)).toMatchSnapshot()
  })

  it('waits for data load', async () => {
    const appStore = new AppStore()
    const componentSetup = (
      <Provider appStore={appStore}>
        <ListPage />
      </Provider>
    )

    const wrapper = mount(componentSetup)
    await when(() => !!appStore.users.length)
    expect(squashHtmlTags(wrapper.html())).toMatchSnapshot()
  })
})
