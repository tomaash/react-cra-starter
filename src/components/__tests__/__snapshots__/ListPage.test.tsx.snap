// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<ListPage /> renders component 1`] = `
"
First Name
Last Name
"
`;

exports[`<ListPage /> waits for data load 1`] = `
"
First Name
Last Name
George
Bluth
Janet
Weaver
Emma
Wong
"
`;
